### PR TITLE
Add support GTM

### PR DESCRIPTION
**Pull request by edglynes**
Initial checkin of gtm support. two new packages, configgtm-v1 and reportsgtm-v1. configgtm supports full API. reportsgtm has limited support for Property IP status and Property|Datacenter Traffic

### CODE REVIEW

File: configgtm-v1/asmap.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+import (
+	"github.com/akamai/AkamaiOPEN-edgegrid-golang/client-v1"
+)
+
+//
+// Handle Operations on gtm asmaps
+// Based on 1.3 schema
+//
+
```
Comment (dshafik) [inline configgtm-v1/asmap.go]: Needs a comment:

`// AsAssignment represents...`

File: configgtm-v1/asmap.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+// AsMap  represents a GTM AsMap   
+type AsMap struct {
+        DefaultDatacenter        *DatacenterBase   `json:"defaultDatacenter"`
+        Assignments              []*AsAssignment   `json:"assignments,omitempty"`
+        Name                     string            `json:"name,omitempty"`
+        Links                    []*Link           `json:"links,omitempty"`
+}
+
+// NewAsMap creates a new asMap
+func NewAsMap(asmapname string) *AsMap {
```
Comment (dshafik) [inline configgtm-v1/asmap.go]: I'd go with `name` as the argument (in every place you use `asmapname`)

File: configgtm-v1/asmap.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	asmap := &AsMap{Name: asmapname}
+        return asmap
+}
+
+// GetAsMap retrieves a asMap with the given name.                     
+func GetAsMap(asmapname, domainname string) (*AsMap, error) {
```
Comment (dshafik) [inline configgtm-v1/asmap.go]: You should use camelCase, so `domainName`

File: configgtm-v1/domain.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+type DomainItem struct {
+        AcgId                    string            `json:"acgId"`
+        LastModified             string            `json:"lastModified"`
+        Links                    []*Link           `json:"links"`
+        Name                     string            `json:"name"`
+        Status                   string            `json:"status"`
+}
+
+func SetHeader(req *http.Request) {
+
```
Comment (dshafik) [inline configgtm-v1/domain.go]: Remove the whitespace here and after the return

File: configgtm-v1/domain.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+type DomainItem struct {
+        AcgId                    string            `json:"acgId"`
+        LastModified             string            `json:"lastModified"`
+        Links                    []*Link           `json:"links"`
+        Name                     string            `json:"name"`
+        Status                   string            `json:"status"`
+}
+
+func SetHeader(req *http.Request) {
```
Comment (dshafik) [inline configgtm-v1/domain.go]: Should this be exported? Or private to the package (e.g. `setHeader()`). You might also consider calling it `setVersionHeader` and allow passing in an argument for the version if you anticipate that changing?

File: configgtm-v1/asmap.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	as := NewAsMap(asmapname)
+	req, err := client.NewRequest(
+		Config,
+		"GET",
+		"/config-gtm/v1/domains/"+domainname+"/as-maps/"+asmapname,
```
Comment (dshafik) [inline configgtm-v1/asmap.go]: consider `fmt.Sprintf()` instead of concatenation

File: configgtm-v1/service.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        // used as trouble shooting hook til logging implemented.
+        debug = false
+ 
+        log.Debugf("Log debug level set to  %v", config.Debug)
+
+}
+
+// Utility func to print http req
+func printHttpRequest(req *http.Request, body bool) {
+
```
Comment (dshafik) [inline configgtm-v1/service.go]: Remove whitespace

File: configgtm-v1/service.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        if !debug {
```
Comment (dshafik) [inline configgtm-v1/service.go]: I would look at using the `sirupsen/logrus` package and the `AKAMAI_LOG` env variable to control output (see [here for how we do it in Akamai CLI](https://github.com/akamai/cli/blob/master/log.go) — I intend to migrate this to the common package at some point for all Golang CLI packages to use also)

File: configgtm-v1/asmap.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+// Instantiate new Assignment struct
+func (as *AsMap) NewAssignment(dcid int, nickname string) *AsAssignment {
+        asAssign := &AsAssignment{}
+        asAssign.DatacenterId = dcid 
+        asAssign.Nickname =nickname
+
+        return asAssign
+
```
Comment (dshafik) [inline configgtm-v1/asmap.go]: whitespace

File: configgtm-v1/asmap.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+		if err != nil {
+			return nil, err
+		}
+
+		return as, nil
+	}
+}
+
+// Instantiate new Assignment struct
+func (as *AsMap) NewAssignment(dcid int, nickname string) *AsAssignment {
```
Comment (dshafik) [inline configgtm-v1/asmap.go]: `dcID`

File: configgtm-v1/asmap.go [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+        stat, err := as.save(domainname)
+        if err != nil {
+                return nil, err
+        }
+        return stat.Status, err
+
+}
+
+// Save AsMap in given domain. Common path for Create and Update.
+func (as *AsMap) save(domainname string) (*AsMapResponse, error) {
```
Comment (dshafik) [inline configgtm-v1/asmap.go]: `Save`

File: configgtm-v1/asmap.go (hunk 1 above)
Comment (edglynes) [reply configgtm-v1/asmap.go]: corrected

File: configgtm-v1/asmap.go (hunk 2 above)
Comment (edglynes) [reply configgtm-v1/asmap.go]: corrected

File: configgtm-v1/asmap.go (hunk 3 above)
Comment (edglynes) [reply configgtm-v1/asmap.go]: corrected

File: configgtm-v1/domain.go (hunk 4 above)
Comment (edglynes) [reply configgtm-v1/domain.go]: corrected. ran go fmt

File: configgtm-v1/domain.go (hunk 5 above)
Comment (edglynes) [reply configgtm-v1/domain.go]: corrected

File: configgtm-v1/service.go (hunk 7 above)
Comment (edglynes) [reply configgtm-v1/service.go]: refactored

File: configgtm-v1/service.go (hunk 8 above)
Comment (edglynes) [reply configgtm-v1/service.go]: refactored to use logrus. Added log.go to edgegrid package. Made SetupLogging generic for anyone to use.

File: configgtm-v1/asmap.go (hunk 10 above)
Comment (edglynes) [reply configgtm-v1/asmap.go]: corrected

File: configgtm-v1/asmap.go (hunk 11 above)
Comment (edglynes) [reply configgtm-v1/asmap.go]: save is intended to be internal and not exposed. Used by Create and Update methods.

**Pull request merged**